src/gc/gc.cpp

```diff
@@ -28868,17 +29104,30 @@ void gc_heap::mark_through_cards_for_segments (card_fn fn, BOOL relocating)
     dprintf(3, ("CMs: %Ix->%Ix", (size_t)beg, (size_t)end));
     size_t total_cards_cleared = 0;
 
+#ifdef FEATURE_CARD_MARKING_STEALING
+    card_marking_enumerator card_mark_enumerator (seg, low, (VOLATILE(uint32_t)*)&card_mark_chunk_index_soh);
+    card_word_end = 0;
+#endif // FEATURE_CARD_MARKING_STEALING
+
     while (1)
     {
         if (card_of(last_object) > card)
         {
             dprintf (3, ("Found %Id cg pointers", cg_pointers_found));
             if (cg_pointers_found == 0)
             {
+#ifdef FEATURE_CARD_MARKING_STEALING
+                uint8_t* co = min(limit, last_object);
+                dprintf(3, (" Clearing cards [%Ix, %Ix[ ", (size_t)card_address(card), (size_t)co));
+                clear_cards(card, card_of(co));
+                n_card_set -= (card_of(co) - card);
+                total_cards_cleared += (card_of(co) - card);
+#else // FEATURE_CARD_MARKING_STEALING
                 dprintf(3,(" Clearing cards [%Ix, %Ix[ ", (size_t)card_address(card), (size_t)last_object));
```

Comment:
could just be rewritten to 

                uint8_t last_object_processed = last_object;
#ifdef FEATURE_CARD_MARKING_STEALING
                last_object_processed = min (limit, last_object);
#endif // FEATURE_CARD_MARKING_STEALING

                clear_cards (card, card_of (last_object_processed));
                n_card_set -= (card_of (last_object_processed) - card);
                total_cards_cleared += (card_of (last_object_processed) - card);

so we don't duplicate as much code

Reply:
and corresponding changes in mark_through_cards_for_large_objects

---
In reply to: [312787476](https://github.com/dotnet/coreclr/pull/25986#discussion_r312787476) [](ancestors = 312787476)

Reply:
Ok, done.